### DISCUSSION TITLE
Wrong number of displacement variables

### DISCUSSION THREAD

**Discussion by RWTHLHK** [idaholab/moose#28043] [Q&A Meshing] [answered]
I generated a quarter cylinder using gmsh and load it into MOOSE. I set the:
[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]
Why do I get wrong number of displacement variables error? Isn't a cylinder a 3D mesh? 
*** ERROR ***
The following error occurred in the Kernel 'TM_all0' of type StressDivergenceTensors.

The number of displacement variables supplied must match the mesh dimension

**Comment by GiudGiud** [comment]
Hello

It should be fine.
Do you mind changing the error message to 
```
    mooseError("The number of displacement variables " + std::to_string(_ndisp) + " supplied must match the mesh dimension " + std::to_string(_mesh.dimension()));
```
and recompiling so we can see which one is counted wrong

Guillaume

**Reply by RWTHLHK** [reply]
Thanks for reply! Now the error message is:
`*** ERROR ***
The following error occurred in the Kernel 'TM_all0' of type StressDivergenceTensors.

The number of displacement variables 3 supplied must match the mesh dimension 2`

**Reply by GiudGiud** [reply]
Do you have a picture of the mesh?

**Reply by RWTHLHK** [reply]
![Screenshot from 2024-07-02 16-17-39](https://github.com/idaholab/moose/assets/43910101/64bef5c3-6e5f-4574-a2b0-a48c930b6fd3)

**Reply by GiudGiud** [reply]
looks 3D to me.
I notice the name is mispelled, can you paste the moose [Mesh] block and the output of `ls` in that directory?

**Reply by RWTHLHK** [reply]
I noticed that and already renamed the file.
The [Mesh] block is:
```
[Mesh]
  [rb]
    type = FileMeshGenerator
    file = 'quarter_cylinder.msh'
  []
[]
```
And ls returns:
```
cleavage.i        cleavage_out.e                                       crack2d_computeCrackedStress_finitestrain_plastic_out.csv  quarter_cylinder.msh
cleavage_out.csv  crack2d_computeCrackedStress_finitestrain_plastic.i  crack2d_computeCrackedStress_finitestrain_plastic_out.e
```

**Reply by GiudGiud** [reply]
no idea then. 
The mesh is made of volumetric elements right? Not just the surface mesh?
Can you edit the gmsh file to see if there is an integrer for dimension somewhere?

**Reply by RWTHLHK** [reply]
Weird. In gmsh it is indeed 3D mesh, but in the output file it is only the surface mesh. I will inspect it into more detail. Thanks for reply!

**Reply by GiudGiud** [reply, marked as answer]
chatgpt has some ideas about this lolol

Copy paste:

If Gmsh is generating only a surface mesh instead of a 3D mesh, you likely need to specify or enable volume meshing. Here are some steps and tips to ensure you get a 3D mesh:

1. **Define the Volume**:
   Ensure that you have defined the volume (3D region) in your geometry. In Gmsh, volumes are defined using the `Volume` keyword.

   ```geo
   // Example: Define a simple box volume
   Point(1) = {0, 0, 0, 1};
   Point(2) = {1, 0, 0, 1};
   Point(3) = {1, 1, 0, 1};
   Point(4) = {0, 1, 0, 1};
   Point(5) = {0, 0, 1, 1};
   Point(6) = {1, 0, 1, 1};
   Point(7) = {1, 1, 1, 1};
   Point(8) = {0, 1, 1, 1};

   Line(1) = {1, 2};
   Line(2) = {2, 3};
   Line(3) = {3, 4};
   Line(4) = {4, 1};
   Line(5) = {1, 5};
   Line(6) = {2, 6};
   Line(7) = {3, 7};
   Line(8) = {4, 8};
   Line(9) = {5, 6};
   Line(10) = {6, 7};
   Line(11) = {7, 8};
   Line(12) = {8, 5};

   Line Loop(13) = {1, 2, 3, 4};
   Plane Surface(14) = {13};

   Line Loop(15) = {5, -9, -8, 12};
   Plane Surface(16) = {15};

   Line Loop(17) = {6, -10, -7, 9};
   Plane Surface(18) = {17};

   Line Loop(19) = {7, -11, -6, 10};
   Plane Surface(20) = {19};

   Line Loop(21) = {8, -12, -5, 11};
   Plane Surface(22) = {21};

   Surface Loop(23) = {14, 16, 18, 20, 22};
   Volume(24) = {23};
   ```

2. **Mesh the Volume**:
   Ensure that you generate the volume mesh. In the Gmsh GUI, you can do this by going to the mesh options and selecting 3D.

   In the `.geo` file, you can specify the meshing command as follows:

   ```geo
   Mesh 3;
   ```

3. **Check Mesh Options**:
   In the Gmsh GUI, you can go to `Tools -> Options -> Mesh -> General` and ensure that the "Mesh dimension" is set to 3.

4. **Ensure All Boundaries are Defined**:
   Make sure that all the surfaces that form the boundary of the volume are properly defined and connected.

5. **Physical Groups** (Optional but recommended):
   Define physical groups for the volume. This helps Gmsh understand what you want to mesh and can improve meshing performance.

   ```geo
   Physical Volume("MyVolume") = {24};
   ```

By following these steps, Gmsh should generate a 3D mesh instead of just a surface mesh. If you still encounter issues, please share your `.geo` file or the specific error messages you are seeing for more tailored assistance.

**Reply by RWTHLHK** [reply]
thanks a lot for your help! Now it works!
